### PR TITLE
Update Module

### PR DESCRIPTION
**Pull request by labkey-ivyc**
(Created for Matt because he is OOO)

### CODE REVIEW

File: api/src/org/labkey/api/module/DefaultModule.java [hunk 1]
```diff
@@ -147,9 +147,11 @@
     private String _buildOS = null;
     private String _buildPath = null;
     private String _sourcePath = null;
+    private boolean _useSourcePath = false;
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/module/DefaultModule.java]: Looks like this was added but never used.

File: api/src/org/labkey/api/premium/PremiumService.java [hunk 2]
```diff
@@ -54,6 +54,12 @@ static void setInstance(PremiumService instance)
 
     boolean isRemoteREnabled();
 
+    default ActionURL getUpdateModuleURL(String module)
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/premium/PremiumService.java:58]: Mark this as @Nullable?

File: api/src/org/labkey/api/module/ModuleLoader.java [hunk 3]
```diff
@@ -699,6 +786,9 @@ else if (!moduleNamePattern.matcher(module.getName()).matches())
                     else
                     {
                         module.setExplodedPath(moduleDir);
+                        if (null != moduleFile && moduleFile.getName().endsWith(".module") && moduleFile.isFile())
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/module/ModuleLoader.java:834]: Make this case-insensitive?

File: query/src/org/labkey/query/olap/ServerManager.java [hunk 4]
```diff
@@ -906,7 +906,7 @@ public void testModuleResourceCache()
             // Make sure the cache retrieves the test OLAP descriptor from this module
 
             Module module = ModuleLoader.getInstance().getModule("query");
-            assertEquals("OLAP descriptors from the query module", 1, MODULE_DESCRIPTOR_CACHE.getResourceMap(module).size());
+            assertEquals("OLAP descriptors from the query module", 4, MODULE_DESCRIPTOR_CACHE.getResourceMap(module).size());
```
Comment (labkey-jeckels) [inline query/src/org/labkey/query/olap/ServerManager.java]: Does this belong here? I don't see any changes to the descriptors

File: core/src/org/labkey/core/admin/AdminController.java [hunk 5]
```diff
@@ -7996,74 +8017,59 @@ private ModulesView(Collection<ModuleContext> contexts, String type, String desc
...
+                                    {
+                                        File zip = module.getZippedPath();
+                                        if (null != zip && zip.getParentFile().equals(externalModulesDir))
+                                            maybeReplaceable = true;
+                                    }
 
-                    out.println("</table>");
+                                    return TR(cl(rowCount.getAndIncrement()%2==0 ? "labkey-alternate-row" : "labkey-row"),
+                                        TD(moduleContext.getName()),
+                                        TD(String.valueOf(moduleContext.getInstalledVersion())),
```
Comment (labkey-jeckels) [inline core/src/org/labkey/core/admin/AdminController.java]: @labkey-adam Do we want to show the schema version here too? I think you were planning to do this, but if it's in a separate FB it won't be easy to merge.

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]

File: api/src/org/labkey/api/module/ModuleLoader.java [hunk 6]
```diff
@@ -298,6 +309,78 @@ public static ServletContext getServletContext()
...
+    // Proxy to teleport ExplodedModuleService from outer ClassLoader to inner
+    static class _Proxy implements InvocationHandler
+    {
+        private final Object delegate;
+        private final Map<String,Method> _methods = new HashMap<>();
+
+        public static ExplodedModuleService newInstance(Object obj)
+        {
+            if (!"org.labkey.bootstrap.LabKeyBootstrapClassLoader".equals(obj.getClass().getName()))
+                return null;
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/module/ModuleLoader.java:306]: Should this error out and/log log?

**Review by labkey-jeckels — CHANGES_REQUESTED** [review]

File: core/src/org/labkey/core/admin/AdminController.java (hunk 5 above)
Comment (labkey-adam) [reply core/src/org/labkey/core/admin/AdminController.java]: Yes, this was already changed to display Release Version and Schema Version on develop. Those changes will need to be merged from develop.

Comment (labkey-adam) [reply core/src/org/labkey/core/admin/AdminController.java]: GitHub is showing that conflict below

File: query/src/org/labkey/query/olap/ServerManager.java (hunk 4 above)
Comment (labkey-matthewb) [reply query/src/org/labkey/query/olap/ServerManager.java]: That change snuck in.  Should take care of itself after next merge from develop.

File: api/src/org/labkey/api/module/ModuleLoader.java (hunk 6 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/module/ModuleLoader.java:306]: When built as a webapp we can run w/o using LabKeyBootstrap classloader.

File: api/src/org/labkey/api/module/ModuleLoader.java (hunk 3 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/module/ModuleLoader.java:834]: I thought about that, but decided to start strict.  Let me know if you think there's a real usage that is not lowercase.

File: api/src/org/labkey/api/module/DefaultModule.java (hunk 1 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/module/DefaultModule.java]: fixed

File: api/src/org/labkey/api/premium/PremiumService.java (hunk 2 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/premium/PremiumService.java:58]: fixed

**Review by labkey-jeckels — APPROVED** [review]

**Review by labkey-adam — APPROVED** [review]

**Pull request merged**